### ISSUE TITLE
older scan plots are not shown on index.html page

### ISSUE THREAD

**Issue by prjemian** [prjemian/spec2nexus#206]
Eric Dufresne <dufresne@anl.gov> emails:

> We're having some issues with the spec plots web service so here is an example from 8-ID-I: 
> https://8id.xray.aps.anl.gov/specplots/8idi/2019/10/stephenson20191016/
> Only one plot gets displayed.

and

> could you restart the specplots web site for 8-ID?
> It stopped working at scan 31 in this directory.
> https://8id.xray.aps.anl.gov/specplots/8ide/2018/11/maass181106/

**Comment by prjemian** [comment]
example data files:

| scan(s) | file |
| ---- | ---- |
| 1-14 | https://8id.xray.aps.anl.gov/specplots/8idi/2019/10/stephenson20191016/stephenson20191016 |
| >31 | https://8id.xray.aps.anl.gov/specplots/8ide/2018/11/maass181106/maass181106 |

**Comment by prjemian** [comment]
Check offline version of spec2nexus:

```
# packages in environment at /home/mintadmin/Apps/anaconda:
#
# Name                    Version                   Build  Channel
spec2nexus                2021.1.3             h39e3cac_0    aps-anl-tag
```

Run offline:  `specplot_gallery -d ./plots stephenson20191016 maass181106`

From file `maass181106`, 90 plots are generated including scans 32 & 33.
![Clipboard02](https://user-images.githubusercontent.com/2279984/67043581-077d9a00-f0f0-11e9-8fc2-2a48a5d39988.jpg)


From file `stephenson20191016`, 15 plots are generated including scans 1-14.
![Clipboard01](https://user-images.githubusercontent.com/2279984/67043458-c1283b00-f0ef-11e9-9bca-85ffdb1575e1.jpg)

**Comment by prjemian** [comment]
@ericmdufresne adds:

> I contacted BCDA last summer about this around June 21.  There was an issue with plotting files with a period in the filename so a filter was written not to include them. A good example are *.log files.

**Comment by prjemian** [comment]
Looking at 8-ID workstation: bronze, in the bluesky environment:

```
(bluesky) 8idiuser@bronze ~/.../startup/spec_support $ conda list spec2nexus
# packages in environment at /home/beams/8IDIUSER/.conda/envs/bluesky:
#
# Name                    Version                   Build  Channel
spec2nexus                2020.0.2                 pypi_0    pypi
```

**Comment by prjemian** [comment]
The base conda environment (Python3 from /APSshare) looks up to date ([latest release is 2021.1.3](https://github.com/prjemian/spec2nexus/releases/tag/2021.1.3)):

```
(bluesky) 8idiuser@bronze ~/.../startup/spec_support $ conda activate base
(base) 8idiuser@bronze ~/.../startup/spec_support $ conda list spec2nexus
# packages in environment at /APSshare/anaconda3/x86_64:
#
# Name                    Version                   Build  Channel
spec2nexus                2021.1.3             h39e3cac_0    aps-anl-tag
(base) 8idiuser@bronze ~/.../startup/spec_support $ which python
/APSshare/anaconda3/x86_64/bin/python
```

**Comment by ericmdufresne** [comment]
Email from Joe Sullivan explaining problem last summer
ED

Date: Mon, 24 Jun 2019 13:55:43 -0500
From: "Sullivan, Joseph" <jpsullivan@anl.gov>
To: "Dufresne, Eric" <dufresne@anl.gov>
Cc: "Walsh, Thomas R." <twalsh@anl.gov>,
     "Narayanan, Suresh" <sureshn@anl.gov>,
     "Strzalka, Joe" <strzalka@anl.gov>
Subject: Re: specplots offline since April 2019

Eric, 

The 8ID spec plots are being updated now. The  process was being asked to
process every file in the spec_data area and some non-data files were
causing it problems. 

I added a file filter to specplots that fix it but it will filter out any
file with a '.' in the name. No data files include a period now so just so
you are aware of that limitation. 

Joe

**Comment by prjemian** [comment]
@ericmdufresne Where is that script running (directory and workstation)?

**Comment by prjemian** [comment]
Also, these files do not have a dot in the name or path (website base URL does not count)

**Comment by ericmdufresne** [comment]
Not sure, Joe S. would know. 
Suresh told me this in 2017:
https://8id.xray.aps.anl.gov/elog/controls/78

Eric

**Comment by prjemian** [comment]
Ok, we'll look on our end.

**Comment by ericmdufresne** [comment]
Weird because this works from Feb. 2019:
https://8id.xray.aps.anl.gov/specplots/8idi/2019/02/zjiang20190201/
Eric

**Comment by prjemian** [comment]
Public website: http://8id.xray.aps.anl.gov/livedata/specplots/

> For production, this VM: s8idapps

**Comment by prjemian** [comment]
@ericmdufresne :  Yes, that is interesting.  Probably the server's log files will show if there were problems with processing.

**Comment by prjemian** [comment]
Also, looking at other directories, lots of missing plots.  Now redirecting attention to the s8idapps server.

**Comment by ericmdufresne** [comment]
Could not get to this public site fyi. Was surprised with http not https
but https://8id.xray.aps.anl.gov/specplots/8idi/2019/ this is live.
Eric

**Comment by prjemian** [comment]
@jsullivan-anl:  Where is the script that calls [`specplot_gallery`](https://spec2nexus.readthedocs.io/en/latest/specplot_gallery.html)?

**Comment by jsullivan-anl** [comment]
/home/joule/WEB8ID/specplots/specplots_8ide
/home/joule/WEB8ID/specplots/specplots_8idg
/home/joule/WEB8ID/specplots/specplots_8idi

Called from a cron job on s8idapps-vm

________________________________

**Comment by prjemian** [comment]
Aha!  As user `web8id`

**Comment by prjemian** [comment]
Probably should move this issue to 8ID repo.  Non suitable now so keep it here.

**Comment by prjemian** [comment]
cron jobs

```
s8idapps-vm% crontab -l
*/5 * * * * /home/joule/WEB8ID/pvWebMonitor/8idi/manage.sh checkup 2>&1 > /dev/null
*/5 * * * * /home/joule/WEB8ID/pvWebMonitor/8ide/manage.sh checkup 2>&1 > /dev/null
*/5 * * * * /home/joule/WEB8ID/pvWebMonitor/8idg/manage.sh checkup 2>&1 > /dev/null
*/5 * * * * /home/joule/WEB8ID/specplots/specplots_8ide 2>&1 > /dev/null
*/5 * * * * /home/joule/WEB8ID/specplots/specplots_8idg 2>&1 > /dev/null
*/5 * * * * /home/joule/WEB8ID/specplots/specplots_8idi 2>&1 > /dev/null
*/1 * * * * /home/joule/WEB8ID/specplots/speclogs_8id 2>&1 > /dev/null
```

**Comment by prjemian** [comment]
Interestingly enough, the plots are actually in the directory and can be viewed individually online, such as:
https://8id.xray.aps.anl.gov/specplots/8idi/2019/10/stephenson20191016/s00002.svg

```
bash-4.2$ ls -lAFgh /home/joule/WEB8ID/www/specplots/8idi/2019/10/stephenson20191016/
total 2.5M
-rw-r--r-- 1 web8id  951 Oct 17 16:50 index.html
-rw-r--r-- 1 web8id  37K Oct 16 12:25 s00001.svg
-rw-r--r-- 1 web8id  43K Oct 16 12:30 s00002.svg
-rw-r--r-- 1 web8id  44K Oct 16 12:35 s00003.svg
-rw-r--r-- 1 web8id  47K Oct 16 12:40 s00004.svg
-rw-r--r-- 1 web8id  41K Oct 16 13:30 s00005.svg
-rw-r--r-- 1 web8id  37K Oct 16 13:30 s00006.svg
-rw-r--r-- 1 web8id  44K Oct 16 15:35 s00007.svg
-rw-r--r-- 1 web8id  38K Oct 16 15:40 s00008.svg
-rw-r--r-- 1 web8id  45K Oct 16 16:40 s00009.svg
-rw-r--r-- 1 web8id  44K Oct 16 16:50 s00010.svg
-rw-r--r-- 1 web8id  44K Oct 16 19:00 s00011.svg
-rw-r--r-- 1 web8id  45K Oct 16 19:45 s00012.svg
-rw-r--r-- 1 web8id  45K Oct 16 19:50 s00013.svg
-rw-r--r-- 1 web8id  43K Oct 17 14:05 s00014.svg
-rw-r--r-- 1 web8id  45K Oct 17 14:30 s00015.svg
-rw-r--r-- 1 web8id  45K Oct 17 15:35 s00016.svg
-rw-r--r-- 1 web8id  45K Oct 17 15:40 s00017.svg
-rw-r--r-- 1 web8id  44K Oct 17 16:50 s00018.svg
-rw-rw-r-- 1 web8id 1.7M Oct 17 16:50 stephenson20191016
```

Next will examine if `index.html` file is written correctly.

**Comment by prjemian** [comment]
No, the `index.html` file is not written correctly to allow those plots to be shown.  It just names the files.  And provides a link to the last one.

```
<html>
  <head>
    <title>SPEC scans from stephenson20191016</title>
    <!-- 
   written by: /APSshare/anaconda3/x86_64/bin/specplot_gallery
   date: 2019-10-17 16:50:20

 -->
  </head>
  <body>
    <h1>SPEC scans from: stephenson20191016</h1>

    spec file: <a href='stephenson20191016'>Oct2019/stephenson20191016</a>
    <br />


    <h2>18 plotted scan(s)</h2>

s00001.svg
s00002.svg
s00003.svg
s00004.svg
s00005.svg
s00006.svg
s00007.svg
s00008.svg
s00009.svg
s00010.svg
s00011.svg
s00012.svg
s00013.svg
s00014.svg
s00015.svg
s00016.svg
s00017.svg
<a href="s00018.svg"><img src="s00018.svg" width="150" height="75" alt="#18: ascan  samx -11 -7  40 0.5"/></a>

    <hr />
    <center><small>page created: 2019-10-17 16:50:20.539741, specplot_gallery documentation: <a href="http://spec2nexus.readthedocs.io/en/latest/specplot_gallery.html">http://spec2nexus.readthedocs.io/en/latest/specplot_gallery.html</a></small></center>

  </body>
</html>
```

**Comment by ericmdufresne** [comment]
Sure is interesting that all the files are in.
Eric

**Comment by prjemian** [comment]
Testing the code that is actually run, the `index.html` file is written properly when writing to a temporary directory but not to the usual directory used by the web server.  This means the software *would* work properly but will not overwrite existing structure.  There is no option to force an overwrite (hint: that might be useful in this instance):

```
(base) bash-4.2$ $PROGRAM -h
usage: specplot_gallery [-h] [-r] [-d DIR] paths [paths ...]

read a list of SPEC data files (or directories) and plot images of all scans

positional arguments:
  paths              SPEC data file name(s) or directory(s) with SPEC data
                     files

optional arguments:
  -h, --help         show this help message and exit
  -r                 sort images from each data file in reverse chronolgical
                     order
  -d DIR, --dir DIR  base directory for output
                     (default:/home/beams10/8IDIUSER/spec_data/2019)
```

**Comment by prjemian** [comment]
What **does** work is to remove the existing directory from the web server and re-create it.

```
/bin/rm -rf $PLOTDIR/2019/10/stephenson20191016
 ${PROGRAM} -r -d ${PLOTDIR} ${FILELIST}
```

and now the plots show up:  https://8id.xray.aps.anl.gov/specplots/8idi/2019/10/stephenson20191016/

**Comment by prjemian** [comment]
I'll check the program logic as to why it might display the file names yet not provide images.  There is an explanation for that *feature*.

The basic repair method right now when this type of thing recurs (plots not showing up for which there is known plottable data):

1. login to the web8id@s8idapps account
1. go to the web server directory (this time: `cd ~/www/specplots/8idi/2019/10`)
1. remove the "*broken*" directory (this time: `/bin/rm -rf ./stephenson20191016`)
1. wait for the cron job to run next -or- run it yourself (this time: `~/specplots/specplots_8idi`)

**Comment by prjemian** [comment]
Both web pages (in the initial report) now show the expected number of plots.  Will check the others in 2019/10

**Comment by prjemian** [comment]
This is more involved.  For a time, the files were viewable and now they are not.  Something else is involved.

**Comment by prjemian** [comment]
Ok, watch the `index.html` files in 2019/10 and see if they change (hypothesis: files are being changed by some other process):

```

-rw-r--r-- 1 web8id 7.3K Oct 17 17:48 2019/10/qzhang20191005/index.html
-rw-r--r-- 1 web8id 9.8K Oct 17 17:48 2019/10/qzhang20191010/index.html
-rw-r--r-- 1 web8id 6.8K Oct 17 17:48 2019/10/rinaldi20191012/index.html
-rw-r--r-- 1 web8id 2.6K Oct 17 18:00 2019/10/stephenson20191016/index.html
```

**Comment by prjemian** [comment]
Here's what that file *should* be:

```
<html>
  <head>
    <title>SPEC scans from stephenson20191016</title>
    <!-- 
   written by: /APSshare/anaconda3/x86_64/bin/specplot_gallery
   date: 2019-10-17 18:00:23

 -->
  </head>
  <body>
    <h1>SPEC scans from: stephenson20191016</h1>

    spec file: <a href='stephenson20191016'>Oct2019/stephenson20191016</a>
    <br />


    <h2>18 plotted scan(s)</h2>

<a href="s00018.svg"><img src="s00018.svg" width="150" height="75" alt="#18: ascan  samx -11 -7  40 0.5"/></a>
<a href="s00017.svg"><img src="s00017.svg" width="150" height="75" alt="#17: ascan  samx -6.53425 -2.53425  40 0.5"/></a>
<a href="s00016.svg"><img src="s00016.svg" width="150" height="75" alt="#16: ascan  samx -6.5 -2.5  40 0.5"/></a>
<a href="s00015.svg"><img src="s00015.svg" width="150" height="75" alt="#15: ascan  samx -2 2  40 0.5"/></a>
<a href="s00014.svg"><img src="s00014.svg" width="150" height="75" alt="#14: ascan  samx -11 -7  40 0.5"/></a>
<a href="s00013.svg"><img src="s00013.svg" width="150" height="75" alt="#13: ascan  samx -11 -7  40 0.5"/></a>
<a href="s00012.svg"><img src="s00012.svg" width="150" height="75" alt="#12: ascan  samx -6.5 -2.5  40 0.5"/></a>
<a href="s00011.svg"><img src="s00011.svg" width="150" height="75" alt="#11: ascan  samx -2 2  40 0.5"/></a>
<a href="s00010.svg"><img src="s00010.svg" width="150" height="75" alt="#10: ascan  samx 2.5 6.5  40 0.5"/></a>
<a href="s00009.svg"><img src="s00009.svg" width="150" height="75" alt="#9: ascan  flyz 2 22  50 0.5"/></a>
<a href="s00008.svg"><img src="s00008.svg" width="150" height="75" alt="#8: ascan  flyz 10.025 20.025  20 0.1"/></a>
<a href="s00007.svg"><img src="s00007.svg" width="150" height="75" alt="#7: ascan  samx -10 10  50 0.2"/></a>
<a href="s00006.svg"><img src="s00006.svg" width="150" height="75" alt="#6: ascan  samx -10 10  50 0.2"/></a>
<a href="s00005.svg"><img src="s00005.svg" width="150" height="75" alt="#5: ascan  samx -4 4  40 0.2"/></a>
<a href="s00004.svg"><img src="s00004.svg" width="150" height="75" alt="#4: timescan 1 0"/></a>
<a href="s00003.svg"><img src="s00003.svg" width="150" height="75" alt="#3: ascan  ti3_x -42.3403 -18.3403  60 1"/></a>
<a href="s00002.svg"><img src="s00002.svg" width="150" height="75" alt="#2: ascan  samx -10 10  40 1"/></a>
<a href="s00001.svg"><img src="s00001.svg" width="150" height="75" alt="#1: ascan  samx 5 5  10 0.1"/></a>

    <hr />
    <center><small>page created: 2019-10-17 18:00:23.725245, specplot_gallery documentation: <a href="http://spec2nexus.readthedocs.io/en/latest/specplot_gallery.html">http://spec2nexus.readthedocs.io/en/latest/specplot_gallery.html</a></small></center>

  </body>
</html>
```

**Comment by prjemian** [comment]
After the next 5 minute interval, the index file is a different size!  Changed from "2.6K" to "951".  We have a problem with another cron job.

```
-rw-r--r-- 1 web8id 7.3K Oct 17 17:48 2019/10/qzhang20191005/index.html
-rw-r--r-- 1 web8id 9.8K Oct 17 17:48 2019/10/qzhang20191010/index.html
-rw-r--r-- 1 web8id 6.8K Oct 17 17:48 2019/10/rinaldi20191012/index.html
-rw-r--r-- 1 web8id  951 Oct 17 18:05 2019/10/stephenson20191016/index.html
```

**Comment by prjemian** [comment]
Here's the list:

```
(base) bash-4.2$ crontab -l
*/5 * * * * /home/joule/WEB8ID/pvWebMonitor/8idi/manage.sh checkup 2>&1 > /dev/null
*/5 * * * * /home/joule/WEB8ID/pvWebMonitor/8ide/manage.sh checkup 2>&1 > /dev/null
*/5 * * * * /home/joule/WEB8ID/pvWebMonitor/8idg/manage.sh checkup 2>&1 > /dev/null
*/5 * * * * /home/joule/WEB8ID/specplots/specplots_8ide 2>&1 > /dev/null
*/5 * * * * /home/joule/WEB8ID/specplots/specplots_8idg 2>&1 > /dev/null
*/5 * * * * /home/joule/WEB8ID/specplots/specplots_8idi 2>&1 > /dev/null
*/1 * * * * /home/joule/WEB8ID/specplots/speclogs_8id 2>&1 > /dev/null
```

Most likely stale file is copied by `/home/joule/WEB8ID/pvWebMonitor/8idi/...`.

**Comment by ericmdufresne** [comment]
Weird, the only active file that could be written is stephenson20191016,
the other files are closed now. Actually, this group scans very rarely.

Eric

**Comment by prjemian** [comment]
Agreed.  It would be useful if the writer put some version or other diagnostic information into the index.html file.  That would be useful now.  

* workstation that ran the software
* username
* name of software
* PID
* version

**Comment by prjemian** [comment]
These would be as HTML comments, not viewable on regular display.

**Comment by prjemian** [comment]
Here's the info in the mtime_cache.json file for the current SPEC data file:

```
    "Oct2019/stephenson20191016": {
        "mtime": 1571354011.379097,
        "size": 1883302
    }
```

With this info, python can check the mtime (last modification time) of the file on disk and compare it against the last time it was opened.  If the mtime is newer AND the file size has increased, python knows the last piece of info that was read the last time (that's at the file size position).  It backs up to the start of that scan and processes from there through the end of the file.  And updates the mtime_cache.json file.  Maybe all that is necessary to force a replot is to change the mtime and size on the file of interest.  Set them both to 1, for example.

**Comment by prjemian** [comment]
It is getting clearer now.  When the gallery program runs and finds existing content (`*.svg` files in this case) it adds them to the new revision of the index.html but (this is the bug) does not add them to the list of items to be shown as plots.  That's a bug.  Finally, something to fix.

**Comment by prjemian** [comment]
That's why this looked OK if the entire directory were removed.  Looked right the first time, subsequent times just remade the last plot so only the last plot image was marked for display.

**Comment by prjemian** [comment]
@ericmdufresne 
Confirmation that this sequence triggers the reported problem.

1. process the SPEC data file (as usual)
1. edit the `mtime_cache.json` file, changing `mtime` and `size` both to `1` (and save)
1. process the SPEC data file again

Observe that only the last plot will be linked.

**Comment by jsullivan-anl** [comment]
All the 8idi 2019 index.html files are broken now. I saw that only the stephenson2019106 index.html was wrong and thought this was fixed (before checking github) so I tried a clean build (deleted the plots directory and the time stamp .json file.  My mistake. Sorry. Another clean build will be needed after the fix is installed.

**Comment by prjemian** [comment]
The fix, v2021.1.4, has been posted online (https://anaconda.org/prjemian/spec2nexus) and is being installed on /APSshare.  An update for the beamlines will be coming shortly.  New data should plot properly.  Existing data files that are not updated will require some manual work to force them to update.
